### ISSUE TITLE
url: tried to send, doesn't seem to work

### ISSUE THREAD

**Issue by dreamcat4** [jreese/znc-push#195]
Hi,
This is my first time using this module. Trying to get `url` to work. Which is happening, but only on the `get` command, which seems to trigger a get request.

Have configured like this:

```
/query *push
set service url
set message_uri http://ga-z170x-g3:84031/growl?title={context}/{nick}&message={message}
set message_length 160
set highlight znc_user
set debug on
```
Then when I do:

```
send message
```

There is no get request generated. Yet there is one on:

```
get
```

I also tried sending a hilight from other user, which didn't work either unfortunately. Nothing happened.

Any ideas?

I compiled with libcurl, but dont think its that, because as we see with the `get` command, although no action is indended there (just to read the variables), yet HTTP get request is being generated.

Here is the full *push module config:

```
18:07:36 <znc_user> get
18:07:36 <...*push> +========================+========================================================================+
18:07:36 <...*push> | Option                 | Value                                                                  |
18:07:36 <...*push> +========================+========================================================================+
18:07:36 <...*push> | away_only              | no                                                                     |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | channel_conditions     | all                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | client_count_less_than | 0                                                                      |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | context                | *                                                                      |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | debug                  | on                                                                     |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | highlight              | znc_user                                                               |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | idle                   | 0                                                                      |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | last_active            | 180                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | last_notification      | 300                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_content        | {message}                                                              |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_length         | 160                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_priority       | 0                                                                      |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_sound          |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_title          | {title}                                                                |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_uri            | http://ga-z170x-g3:8403/growl?title={context}/{nick}&message={message} |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_uri_post       | no                                                                     |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | message_uri_title      |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | network_blacklist      |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | nick_blacklist         |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | proxy                  |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | proxy_ssl_verify       | yes                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | query_conditions       | all                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | replied                | yes                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | secret                 |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | service                | url                                                                    |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | target                 |                                                                        |
18:07:36 <...*push> +------------------------+------------------------------------------------------------------------+
18:07:36 <...*push> | username               |                                                                        |
18:07:36 <...*push> +========================+========================================================================+
```

**Comment by dreamcat4** [comment]
Oh,
and here is the debug infos from `send message`... which does nothing (no HTTP get recieved that time)

```
18:12:57 <znc_user> send message
18:12:57 <...*push> service: url
18:12:57 <...*push> service_host: ga-z170x-g3
18:12:57 <...*push> service_url: /growl
18:12:57 <...*push> service_auth:
18:12:57 <...*push> use_port: 8403
18:12:57 <...*push> use_ssl: 0
18:12:57 <...*push> use_post: 0
18:12:57 <...*push> using libcurl
18:12:57 <...*push> Ok
```

**Comment by dreamcat4** [comment]
OK so I have tweaked some other setting and now a channel hilight trigger an event, doing the same thing as `send message`

```
18:24:18 <*push> service: url
18:24:18 <*push> service_host: ga-z170x-g3
18:24:18 <*push> service_url: /growl
18:24:18 <*push> service_auth:
18:24:18 <*push> use_port: 8403
18:24:18 <*push> use_ssl: 0
18:24:18 <*push> use_post: 0
18:24:18 <*push> using libcurl
```

Sadly, still no http get received at other end.

**Comment by jreese** [comment]
For comparison, can you try to run `curl http://ga-z170x-g3:8403/growl?title=Foo&message=Bar` from a terminal on the same machine that's running ZNC, and see if that works?  I suspect a DNS issue or similar due to "ga-z170x-g3" not having a TLD, and would like to see how curl behaves from the shell.

**Comment by dreamcat4** [comment]
@jreese Yeah I am happy to ry. But no, that works all fine.

It is like I tried to explain above ^^:

if i just type the word `get` in the `*push` window, it DOES WORK. So that proves your module can send a message / is actually capable of sending a real HTTP get request to the target uri. It appears / can see at the other end just fine.

The problem here is that just doesnt happen for any other ways I have tried (=2 ways). While the debug message is printed, but no error is being reported by it. Yet the result isnt good, and not same as when triggered by your module's `get` command.

Furthermore: If we look at the debug output, we can see the last message is `using libcurl`. And there is nothing more, because the `Ok` at very end is just some boilerplate and not a real confirmater (for example probably generated by your module's command handler function). As all other commands also say `Ok`.

Perhaps we might expect more debug line(s) after that final line `using libcurl`? But for me, there is none.

Module was compiled like this:

```sh
FROM ubuntu:xenial-20160125 # ubuntu 16.04 base image

# Supportive pkgs
RUN apt-get update -qqy && DEBIAN_FRONTEND=noninteractive apt-get install -y \
    wget curl git sudo nano less man && $_apt_clean

# ZNC - install build time dependancies
RUN apt-get update -qqy && DEBIAN_FRONTEND=noninteractive apt-get install -y \
    znc znc-dev ca-certificates libcurl4-openssl-dev make && $_apt_clean


# Push and http get nofitications for mobile etc
RUN git clone https://github.com/jreese/znc-push.git \
 && cd znc-push && make curl=yes \
 && ls -lsa && cp push.so /build/znc-modules/out
```

**Comment by dreamcat4** [comment]
Hi again,

Since yesterday I've had the chance to recompile, so this time removed the `curl=yes` flag and rebuilt without the curl support... Now without curl, the:

* `send message` works fine
* in-channel nick hilight - works fine
* `get` command still sends an HTTP get (maybe by accident)

Maybe it is something todo with the curl library? I just dont know. Its difficult for me to know without some specific instruction / advice to do anything different.

The README does not mention which specific *runtime* curl dependancy needs to be installed. It's compiled statically? README only mentions the *build-time* curl pkg depenancy.

Which for this build was selected as `libcurl4-openssl-dev`. But had to use `libcurl3` for the runtime environment. Since there is no `libcurl4` pkg in ubuntu. Nor visa-versa any alternative `libcurl3-dev` pkg neither.

Anyway here is what is currently available on the official ubuntu 16.04 apt repos:

```
$ apt search libcurl
libcurl-ocaml/xenial 0.5.3-2build5 amd64
  OCaml curl bindings (Runtime Library)

libcurl-ocaml-dev/xenial 0.5.3-2build5 amd64
  OCaml libcurl bindings (Development package)

libcurl3/xenial,now 7.47.0-1ubuntu2 amd64 [installed]
  easy-to-use client-side URL transfer library (OpenSSL flavour)

libcurl3-dbg/xenial 7.47.0-1ubuntu2 amd64
  debugging symbols for libcurl (OpenSSL, GnuTLS and NSS flavours)

libcurl3-gnutls/xenial 7.47.0-1ubuntu2 amd64 [upgradable from: 7.47.0-1ubuntu1]
  easy-to-use client-side URL transfer library (GnuTLS flavour)

libcurl3-nss/xenial 7.47.0-1ubuntu2 amd64
  easy-to-use client-side URL transfer library (NSS flavour)

libcurl4-doc/xenial 7.47.0-1ubuntu2 all
  documentation for libcurl

libcurl4-gnutls-dev/xenial 7.47.0-1ubuntu2 amd64
  development files and documentation for libcurl (GnuTLS flavour)

libcurl4-nss-dev/xenial 7.47.0-1ubuntu2 amd64
  development files and documentation for libcurl (NSS flavour)

libcurl4-openssl-dev/xenial 7.47.0-1ubuntu2 amd64
  development files and documentation for libcurl (OpenSSL flavour)

libcurlpp-dev/xenial 0.7.3-6 amd64
  c++ wrapper for libcurl (development files)

libcurlpp0/xenial 0.7.3-6 amd64
  c++ wrapper for libcurl
```

As can be seen, both those selected pkgs say `7.47`. So I'm genuinely stumped by it all. I.e. which should be the correct deps pkgs. It being ubuntu and all.

The curl (standalone program's) deps:

```
$ apt show curl
Package: curl
Version: 7.47.0-1ubuntu2
Depends: libc6 (>= 2.17), libcurl3-gnutls (= 7.47.0-1ubuntu2), zlib1g (>= 1:1.1.4)
```

**Comment by faxm0dem** [comment]
I have the same issue. I set up a dummy web server and I found that everything after the question mark (the GET query parameters) is plainly and simply discarded. I tried to escape the question mark to no avail.
